Fixed: Scroll jumps when closing dialogs

engineering-this opened the pull request:
## What is the purpose of this pull request?

Bug fix + new feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [X] Unit tests
- [X] Manual tests

## What changes did you make?

New feature part:

- Both `editor.focus` and `editable.focus` have now optional param object with property `preventScroll`. If it's set to true focusing won't affect scroll.
- Using the native focus with that flag for Chrome and Safari.
- Other browsers uses simple fallback: storing scroll position of any scrollable element and restoring it after calling native focus.
- Edge: detecting of scrollable element is not possible until [14721015](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/14721015/) is fixed. 

Bugfix part:

- When the dialog is opened (by calling dialog command) it focuses the editor with `{ preventScroll: true }`.
- When the `dialog.hide` is called it focuses the editor with same flag.

And about mobiles: they are ignored in manuals and I don't think we can handle this easily within this ticket.

Note: first case wasn't reported, but I noticed that opening dialog on blurred editor will also trigger focus with scroll, so added fix for that too.

Closes #748 
Closes #895
Closes #1429
Closes #2420

jacekbogdanski commented:
What about an additional flag like `editable.focus( scroll = true )`? I think it would be nice to have both options available.

engineering-this commented:
Thanks @jacekbogdanski for your input. I guess this is best option.

engineering-this commented:
Edited PR message to reflect reworked solution.

Comandeer commented on tests/plugins/dialog/manual/pagescroll.html:
```diff
@@ -0,0 +1,8 @@
+<div style="min-height: 100vh"></div>
+<textarea id="editor"></textarea>
+<script>
+	if ( CKEDITOR.env.ie || bender.tools.env.mobile ) {
```
Is there any other reason to ignore manual tests in IEs than lack of support for `vh` unit? If no, then this issue could be fixed with using large `px` value for `height`.

Comandeer replied:
Still ignored…

engineering-this commented on core/editable.js:
```diff
@@ -81,29 +96,72 @@
+										} );
+
+										// IE and Edge scroll asynchronously.
+										if ( CKEDITOR.env.ie ) {
```
I could drop IIFE by moving this part to separe forEach loop. But I'm not sure if it's better to loop two times, or just use same loop, but with IIFE for function scoping.

Comandeer replied:
I'm for keeping IIFE.

engineering-this commented:
Fix now works with IE and Edge and they aren't ignored in tests any more.

It looks like `setActive` which was called on IE instead of `focus` was scrolling page even if our comments said it didn't.

Comandeer commented on core/editable.js:
```diff
@@ -69,7 +84,7 @@
 					active = this.editor._.previousActive || this.getDocument().getActive();
 					if ( this.contains( active ) ) {
-						active.focus();
+						active.focus( focusOptions );
```
This invocation does not work as intended. `active` is an instance of `CKEDITOR.dom.element`, not an editable. In such case, the signature of the `focus` method is different. In fact passing this parameter causes `focus` to be called asynchronously, as is treated as `defer` attribute. Please see: https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_dom_element.html#method-focus

That creates the tricky situation when the new fix is not working with nested editables (e.g. captions in `image2` or `easyimage` plugins), as viewport is always scrolled to these elements when inline editable is used (it's not present in classic, `iframe` based editor, as in such situation the `iframe` itself is a viewport, so only the content of `iframe` is scrolled, not the whole page).

I don't see any really sensible solution to that. We can overload `CKEDITOR.dom.element#focus` method to treat the fist parameter as `defer` one if value other than object is passed and as `focusOptions` if object is passed (in the second case we could make `defer` one of options, as probably there will be cases when both of the behaviours will be needed). It'd result in moving most of the logic from `CKEDITOR.editable` to `CKEDITOR.dom.element`. After that changes `CKEDITOR.editable#focus` will be using `CKEDITOR.dom.element#focus`, only adding some new behaviour.

However this change is not trivial anymore and requires much more tests. WDYT @engineering-this, @mlewand?

engineering-this commented:
I moved logic to `CKEDITOR.dom.element.prototype.focus`. So far it looks to be working fine. I kept `focusOption` to overload element method. So now we have two possible options `preventScroll` and `defer`.

Comandeer commented on core/editor.js:
````diff
@@ -1305,9 +1305,25 @@
+		 * editor.focus(); // Focuses without preventing scroll.
+		 * ```
+		 *
+		 * @param {Object} [focusOptions] Focus options since **4.11.0**.
````
We describe `focusOptions` in three different places, everywhere the same way.

It seems that we need… [DU-DU-DUUUUUUM](https://www.youtube.com/watch?v=UgwSSAzxcs8) dedicated type for it.

Comandeer commented on tests/core/editable/manual/focus.html:
```diff
@@ -0,0 +1,38 @@
+<div style="height:8000px"></div>
+
+<div id="inline" contenteditable="true">
+	<figure class="easyimage">
```
Using widget as the only content inside the editor makes testing if editor is focused more difficult. The most obvious way (start typing) is not working anymore.

Comandeer commented on core/editable.js:
```diff
@@ -81,34 +99,13 @@
-					if ( !CKEDITOR.env.ie )
-						throw e;
-				}
+				CKEDITOR.dom.element.prototype.focus.call( this, focusOptions );
```
IMO it's better (_more elegant_) to call this class base class method:

```javascript
CKEDITOR.editable.baseProto.focus.call( this, focusOptions );
```

engineering-this commented on core/dom/element.js:
```diff
@@ -419,26 +419,135 @@ CKEDITOR.dom.element.clearMarkers = function( database, element, removeFromDatab
 				try {
-					this.$.focus();
-				} catch ( e ) {}
+					var isScrollLockFallbackNeeded = !CKEDITOR.env.webkit;
```
As we have talked, I'm implementing detection for `focusOptions` in test, so it doesn't affect editor.

engineering-this commented on core/editable.js:
```diff
@@ -59,8 +59,26 @@
+			 *
+			 * @param {Object} [focusOptions] Focus options since **4.11.0**.
+			 * @param {Boolean} [focusOptions.preventScroll] Whenever after focus scroll should be prevented.
+			 * @param {Boolean} [focusOptions.defer] Whether to asynchronously defer the
```
Actually I don't think we need it, but as this method calls element.focus with given focus options we would have to explicitly `delete focusOptions.defer`. Otherwise it would work as an undocumented feature. I'm for keeping it. OTOH as I'm moving `focusOptions` docs to dedicated type to keep it consistent - each focus method should allow same `focusOptios` type holding same properties.

engineering-this commented on core/dom/element.js:
```diff
@@ -419,26 +419,135 @@ CKEDITOR.dom.element.clearMarkers = function( database, element, removeFromDatab
+										}
+									} )();
+								}
+								if ( element.getParent() ) {
```
Just to clarify:

>That's not true for CKEditor as we want to keep scrolling position inside the editable without preventing the scrolling of the viewport (see https://dev.ckeditor.com/ticket/14825 ).

I assume you have meant 'we want to keep _scroll_ position...'. In other words: always prevent scrolling inside editor.

engineering-this commented on core/editable.js:
```diff
@@ -81,34 +99,13 @@
+				CKEDITOR.dom.element.prototype.focus.call( this, focusOptions );
 
-				// Remedy if Safari doens't applies focus properly. (https://dev.ckeditor.com/ticket/279)
+				// Remedy if Safari doesn't applies focus properly. (https://dev.ckeditor.com/ticket/279)
```
It comes from https://github.com/ckeditor/ckeditor-dev/commit/0f19031124d5e496471e93aca92b899fa89f9dc3 there were no tests included, but nothing looks broken after removing it, and editor is focused after closing floatpanel.

engineering-this commented on core/dom/element.js:
```diff
@@ -419,28 +419,117 @@ CKEDITOR.dom.element.clearMarkers = function( database, element, removeFromDatab
+
+					if ( isScrollLockFallbackNeeded && focusOptions && focusOptions.preventScroll ) {
+						scrollables = storeScrollPosition( element );
+					} else if ( !isInline && CKEDITOR.env.chrome && ( !focusOptions || !focusOptions.preventScroll ) ) {
```
To fix this Chrome bug we need to restore scroll position on editable parent, so general approach to restore  scroll position only in focused element wouldn't work in that case.

engineering-this commented:
- Fixed case when editor scrolled in IE8 when focused and in Chrome for inline editors.
- Simplified manual test.
- Refactored code a bit as.

I didn't rebase it yet, I will do so, when `major` branch is ready for commits.

Comandeer commented:
Rebased onto latest `major`.

engineering-this commented on core/dom/element.js:
```diff
@@ -419,28 +419,127 @@ CKEDITOR.dom.element.clearMarkers = function( database, element, removeFromDatab
+							break;
+						}
+
+						if ( !element.getParent() ) {
```
```
element = element.getParent();

if ( !element ) {
	element = element.getWindow().getFrame();
}
```
This wouldn't work, as you are overwritting element with null and then trying to access it's properties.

The second one is perfect tho. I was so afraid of this typeError I didn't realise the simplest solution 🤦‍♂️

Comandeer commented on core/dom/element.js:
```diff
@@ -419,28 +419,127 @@ CKEDITOR.dom.element.clearMarkers = function( database, element, removeFromDatab
+
+				function restoreScrollAndCleanScrollables( scrollables ) {
+					restoreScrollPosition( scrollables );
+					scrollables = [];
```
This line does not clean `scrollables` variable from outer scope, it's just change object referenced by _local_ `scrollables` variable. To really clean `scrollables`, `scrollables.length = 0` could be used.

engineering-this commented:
`tests/core/dom/element/focusoptions`

I fixed this test on Edge, although there is still small issue. When you refresh page with test it fails, but for some reason browser doesn't focus element when refreshing. I don't think we should worry, as test passes when run from dashboard, and when run solo. It's isn't supposed to detect any regressions.

engineering-this commented:
>I wonder about fixing the issue with flickering. Maybe temporal overflow: hidden for scrollable containers would do the trick?

@Comandeer This won't work. `overflow: hidden` hides the scrollbars and content, but doesn't prevent scrolling. Browser/JS can scroll content anyway.

See example:
https://codepen.io/anon/pen/BvmYmV

engineering-this commented:
It is possible to focus element without scroll and even scroll event happening, but this requires to change target elements positioning to `fixed` before focus, and restore positioning after focus.

However in Ege/IE scroll is asynchronous so restoring styles should be wrapped in `setTimeout`, and this leads to another visual flickering, now instead of scroll we have flying elements. To prevent that we should get elements actual position and size, then place next to it some dummy element that will take same space, and add positioning to our focused element so it is rendered in same place when it has fixed position.

But still there is scroll to an iframe, so same logic should be applied on it. 

One would say: why do we care about positioning of elements that out of view?
Actually they can be visible partially and page still would be scrolled on focus. Also some visible container might collapse when it's focusable child is changed to absolute positioning.

I can't think of other options, so I guess for now we will have to live with Edge/IE flickering.

hissy commented:
Is this issue still open?

f1ames commented:
Hello @hissy,
yes, this PR is still work in progress. However, we have shift our focus to similar dialog related fix which is https://github.com/ckeditor/ckeditor-dev/pull/2413 - it may solves some of issues mentioned here, but we have to check it. You may watch https://github.com/ckeditor/ckeditor-dev/pull/2413 to stay updated.

hissy commented:
Thank you, I'd like to check that issue too.
